### ISSUE TITLE
Consolidate a mixture of mutating and functional style in AtomCollection::post_process

### ISSUE THREAD

**Issue by torhovland** [tweag/topiary#96]
The function currently does this:

```rust
        self.put_before(Atom::IndentEnd, Atom::Space, &[]);
        self.put_before(Atom::Hardline, Atom::Blankline, &[Atom::Space]);
        self.put_before(Atom::IndentStart, Atom::Space, &[]);
        self.put_before(Atom::IndentStart, Atom::Hardline, &[Atom::Space]);
        self.put_before(Atom::IndentEnd, Atom::Hardline, &[Atom::Space]);
        self.atoms = self.clean_up_consecutive(Atom::Space, &[]);
        self.atoms = self.clean_up_consecutive(Atom::Hardline, &[Atom::Space]);
        self.atoms = self.trim_following(Atom::Hardline, Atom::Space);
        self.atoms = self.trim_following(Atom::Blankline, Atom::Space);
        self.remove_before(Atom::Hardline, Atom::Blankline, &[Atom::Space]);
        self.ensure_final_hardline();
```

While this works, and doesn't seem to matter much in terms of performance, the mixing of a mutating and a functional style is inconsistent and should be consolidated. The most straightforward way to do so is probably to make `clean_up_consecutive` and `trim_following` mutating.

**Comment by torhovland** [comment]
It's possible that you will see a performance gain on a big input file. In any case, I like how you've simplified the various post-processing steps into one reordering pattern. The old code felt hacky to me.

**Comment by evertedsphere** [comment]
Closed by #102, I believe.
